src/k8s/cmd/k8sd/k8sd_cluster_recover.go

```diff
@@ -0,0 +1,303 @@
+	//       "core_cluster_members" table, executed when k8sd starts.
+	// * rewrite cluster.yaml
+	// * create a recovery tarball of the k8sd database dir and store it
+	//   in the state dir.
+	tarballPath, err := m.RecoverFromQuorumLoss(newMembers)
+	if err != nil {
+		return fmt.Errorf("K8sd recovery failed: %w", err)
+	}
+
+	fmt.Printf("Cluster changes applied; new database state saved to %s\n\n", tarballPath)
+	fmt.Printf("*Before* starting any cluster member, copy %s to %s on all remaining cluster members.\n\n", tarballPath, tarballPath)
+	fmt.Printf("k8sd will load this file during startup.\n")
+	return nil;
+}
+
+func recoverK8sDqlite() error {
+	k8sDqliteStateDir := clusterRecoverOpts.K8sDqliteStateDir
+
+	clusterYamlPath := path.Join(k8sDqliteStateDir, "cluster.yaml")
+	oldMembersYaml, err := os.ReadFile(clusterYamlPath)
+	if err != nil {
+		return err
+	}
+	clusterYamlCommentHeader := fmt.Sprintf("# k8s-dqlite cluster configuration\n# (%s)\n", clusterYamlPath)
+	clusterYamlContent, err := shared.TextEditor(
+		"",
+		slices.Concat(
+			[]byte(clusterYamlCommentHeader),
+			[]byte(clusterYamlRecoveryComment),
+			oldMembersYaml))
+	if err != nil {
+		return err
+	}
+	err = removeYamlCommentsFromFile(clusterYamlPath)
+	if err != nil {
+		return err
+	}
+
+	infoYamlPath := path.Join(k8sDqliteStateDir, "info.yaml")
+	infoYamlCommentHeader := fmt.Sprintf("#k8s-dqlite info.yaml\n# (%s)\n", infoYamlPath)
+	_, err = shared.TextEditor(
+		"",
+		slices.Concat(
+			[]byte(infoYamlCommentHeader),
+			[]byte(infoYamlRecoveryComment),
+			yamlCommentLines(removeYamlComments(clusterYamlContent))))
+	if err != nil {
+		return err
+	}
+	err = removeYamlCommentsFromFile(infoYamlPath)
+	if err != nil {
+		return err
+	}
+
+	newMembers := []cluster.DqliteMember{}
+	err = yaml.Unmarshal(clusterYamlContent, &newMembers)
+	if err != nil {
+		return err
+	}
+
+	nodeInfo := make([]dqlite.NodeInfo, 0, len(newMembers))
+	for _, member := range newMembers {
+		info, err := member.NodeInfo()
+		if err != nil {
+			return err
+		}
+		nodeInfo = append(nodeInfo, *info)
+	}
+
+	err = dqlite.ReconfigureMembershipExt(k8sDqliteStateDir, nodeInfo)
+	if err != nil {
+		return fmt.Errorf("k8s-dqlite recovery failed: %w.", err)
+	}
+
+	// TODO: steps performed by the microcluster lib for k8sd that could also
+	// apply to k8s-dqlite:
+	//   * database backups
+	//     * a tarball of the database dir stored in the state dir
+	//     * for some reason, two tarballs are creaed: one before the recovery and another one after the recovery.
+	//        https://github.com/canonical/microcluster/blob/9da23064cf0ad86768e05db4695a9d239bb717c1/internal/recover/recover.go#L135
+	//        https://github.com/canonical/microcluster/blob/9da23064cf0ad86768e05db4695a9d239bb717c1/internal/recover/recover.go#L172
+	//     * Questions:
+	//        * why does microcloud create two tarballs of the database directory?
+	//          * do we need to do the same?
```

Comment:
Added the same for k8s-dqlite. One difference is that k8sd automatically loads the tarball, in case of k8s-dqlite the user will have to extract it after copying it over to the other nodes.